lib/puppet/pops/types/p_encrypted_type.rb

```diff
@@ -0,0 +1,354 @@
+
+      if ciphers.length == 1 || Puppet[:accepted_ciphers].empty?
+        # Produce the first value in ciphers that is accepted - or nil if non were
+        # Since all ciphers are accepted there is no defined order, so  use first in list of ciphers
+        (ciphers & pruned_ciphers)[0]
+      else
+        # Since there is a preferred order, the pruned[0] contains the preferred cipher name
+        # (acceptable and available orders) are in order of preference when accepted_ciphers is defined
+        pruned_ciphers[0]
+      end
+    end
+
+    # Decrypts this `Encrypted` and returns a `Sensitive` instance with the decrypted value
+    #
+    # @param scope [Puppet::Parser::Scope] Scope is used to find data types for serialization
+    # @param host [Puppet::SSL::Host] The host to decrypt for - defaults to localhost
+    # @return [Puppet::Pops::Types::PSensitiveType::Sensitive] with the decrypted value
+    # @api public
+    #
+    def decrypt(scope, host = Puppet::SSL::Host.localhost)
+      key = host.key
+      raise Puppet::DecryptionError.new(_("No private key available for %{host}") % {host: host.name }) unless key
+      key = key.content
+
+      # decrypt the random key encrypted with receiver's public key
+      begin
+        aes_key = key.private_decrypt(@encrypted_key.binary_buffer)
+      rescue OpenSSL::PKey::RSAError => e
+        raise Puppet::DecryptionError.new(_("Decryption failed (probably using wrong host), caused by: %{message}") % {message: e.message})
+      end
+
+      _serializer, cipher = @format.split(/,/)
+      # Decrypt using the given cipher, and the random key
+      # Use the 16 first byte as IV (they are in clear text)
+      # Decrypt the rest with the random key and the IV
+      #
+      aes_decrypt = OpenSSL::Cipher.new(cipher).decrypt
+      aes_decrypt.key = aes_key
+      aes_decrypt.iv, crypt_part = iv_split(@crypt)
+      data = aes_decrypt.update(crypt_part) << aes_decrypt.final
+
+      # Decrypt (same way) the fingerprint. While not really a secret there is no need
+      # to make it easy to understand who the recipient of an encryption is.
+      #
+      aes_decrypt.reset
+      aes_decrypt.iv, fingerprint_part = iv_split(@encrypted_fingerprint)
+      fingerprint = aes_decrypt.update(fingerprint_part) << aes_decrypt.final
+
+      # Check if this was for "this recipient"
+      #
+      unless fingerprint == host.certificate.fingerprint
+        raise Puppet::DecryptionError.new(_("Decryption failed, the Encrypted is not encrypted for given host"))
+      end
+
+      # The data is a Pcore serialization so it must be deserialized
+      clear = deserialize(data, scope)
+      sensitive = Puppet::Pops::Types::PSensitiveType::Sensitive
+
+      # Ensure result is a Sensitive
+      clear.is_a?(sensitive) ? clear : sensitive.new(clear)
+    end
+
+    # @api private
+    def iv_split(bin)
+      buffer = bin.binary_buffer
+      [buffer[0..15], buffer[16..-1]]
```

Comment:
This assumes we're using 128-bit symmetric algorithm. The encryptor should send the IV and the encrypted value separately, so the decryptor doesn't have to guess (as the IV could contain the ASCII char for `,`). Also some ciphers don't require an IV, so it should be possible to indicate that the "encryptor didn't use an iv".